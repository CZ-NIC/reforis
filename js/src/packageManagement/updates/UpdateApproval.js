/*
 * Copyright (C) 2019-2021 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";
import ReactTooltip from "react-tooltip";

import {
    useAPIPost,
    Button,
    Spinner,
    useAlert,
    API_STATE,
    Alert,
    ALERT_TYPES,
    toLocaleDateString,
    ForisURLs,
} from "foris";
import moment from "moment";
import API_URLs from "common/API";

UpdateApproval.propTypes = {
    update: PropTypes.object.isRequired,
    onSuccess: PropTypes.func.isRequired,
    delay: PropTypes.number,
};

export default function UpdateApproval({ update, onSuccess, delay }) {
    const [setAlert, dismissAlert] = useAlert();

    const [postState, post] = useAPIPost(API_URLs.approvals);
    // Execute callback when resolution is successful
    useEffect(() => {
        if (postState.state === API_STATE.ERROR) {
            setAlert(_("Cannot install updates."));
        } else if (postState.state === API_STATE.SUCCESS) {
            onSuccess();
            setAlert(
                _("Updates will be installed shortly."),
                ALERT_TYPES.SUCCESS
            );
        }
    }, [postState, onSuccess, setAlert]);

    function resolveUpdate(solution) {
        dismissAlert();
        post({ data: { hash: update.hash, solution } });
    }

    if (postState.state === API_STATE.SENDING) {
        return <Spinner className="text-center" />;
    }

    const packagesNumber =
        update.approvable &&
        babel.format(
            ngettext(
                "There is %d package to be updated.",
                "There are %d packages to be updated.",
                update.plan.length
            ),
            update.plan.length
        );
    const details = _(
        'See <a data-toggle="collapse" href="#plan-wrapper" role="button" aria-expanded="false" aria-controls="plan-wrapper">details</a>'
    );
    const delayedTime = moment(update.time).add(delay, "hour");
    return (
        <>
            {update.approvable ? (
                <>
                    {delay && (
                        <Alert type={ALERT_TYPES.SUCCESS}>
                            <span>
                                {_(
                                    `These updates are not going to be 
                                    automatically installed before 
                                    ${toLocaleDateString(delayedTime)}.`
                                )}
                            </span>
                            <i
                                className="fas fa-question-circle ml-1 help"
                                data-for="delay-tooltip"
                                data-tip={_(
                                    `If you don't want the updates to be installed at all, go to
                                    <a href="${ForisURLs.packageManagement.updateSettings}">Update Settings</a> 
                                    and choose the option <b>Update approval needed</b>.`
                                )}
                                data-event="click focus"
                                data-html
                            />
                            <ReactTooltip
                                effect="solid"
                                globalEventOff="click"
                                clickable
                                id="delay-tooltip"
                            />
                        </Alert>
                    )}
                    <h3>
                        {babel.format(
                            _("Approve Update From %s"),
                            toLocaleDateString(update.time)
                        )}
                    </h3>
                    <p
                        dangerouslySetInnerHTML={{
                            __html: `${packagesNumber} ${details}`,
                        }}
                    />
                    <div
                        className="collapse"
                        id="plan-wrapper"
                        data-testid="plan-wrapper"
                    >
                        <Plan plan={update.plan} />
                    </div>
                    <div className="text-right">
                        <Button
                            className="btn-primary"
                            onClick={() => resolveUpdate("grant")}
                            forisFormSize
                        >
                            {_("Install now")}
                        </Button>
                    </div>
                </>
            ) : (
                <p className="text-center text-muted">
                    {_("There are no updates awaiting your approval.")}
                </p>
            )}
        </>
    );
}

Plan.propTypes = {
    plan: PropTypes.arrayOf(PropTypes.object).isRequired,
};

function Plan({ plan }) {
    return (
        <div className="table-responsive">
            <table className="table table-hover">
                <thead className="thead-light">
                    <tr>
                        <th scope="col">{_("Action")}</th>
                        <th scope="col">{_("Name")}</th>
                        <th scope="col">{_("Current")}</th>
                        <th scope="col">{_("New")}</th>
                    </tr>
                </thead>
                <tbody>
                    {plan.map((item) => (
                        <PlanItem key={item.name} {...item} />
                    ))}
                </tbody>
            </table>
        </div>
    );
}

PlanItem.propTypes = {
    name: PropTypes.string.isRequired,
    op: PropTypes.string.isRequired,
    new_ver: PropTypes.string,
    cur_ver: PropTypes.string,
};

function PlanItem({ name, op, new_ver, cur_ver }) {
    return (
        <tr>
            <td>{op}</td>
            <td>{name}</td>
            <td>{cur_ver}</td>
            <td>{new_ver}</td>
        </tr>
    );
}
