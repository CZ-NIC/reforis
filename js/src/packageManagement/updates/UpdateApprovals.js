/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import {
    useAPIGet, useAPIPost, Button, Spinner,
} from "foris";
import API_URLs from "common/API";
import toLocaleDateString from "utils/localeDate";

import isUpdateAvailable from "./utils";

export default function UpdateApprovals() {
    const [getState, get] = useAPIGet(API_URLs.approvals);
    const approval = getState.data;
    useEffect(() => {
        get();
    }, [get]);

    const [postState, post] = useAPIPost(API_URLs.approvals);
    useEffect(() => {
        get();
    }, [get, postState.data]);

    function postHandler(solution) {
        return (e) => {
            e.preventDefault();
            post({ hash: approval.hash, solution });
        };
    }

    if (getState.isLoading || postState.isSending) return <Spinner className="row justify-content-center" />;

    if (!isUpdateAvailable(approval)) {
        return <p>{_("There are no updates awaiting your approval.")}</p>;
    }

    const packagesNumber = babel.format(
        ngettext(
            "There is %d package to be updated.",
            "There are %d packages to be updated.",
            approval.plan.length,
        ),
        approval.plan.length,
    );
    const details = _(
        "See <a data-toggle=\"collapse\" href=\"#plan-wrapper\" role=\"button\" aria-expanded=\"false\" aria-controls=\"plan-wrapper\">details</a>",
    );

    const buttonMargin = { marginBottom: "1rem" };

    return (
        <>
            <h3>{babel.format(_("Approve update from %s"), toLocaleDateString(approval.time))}</h3>
            <p dangerouslySetInnerHTML={{ __html: `${packagesNumber} ${details}` }} />
            <div className="collapse" id="plan-wrapper" data-testid="plan-wrapper">
                <Plan plan={approval.plan} />
            </div>
            <Button
                style={buttonMargin}
                className="btn btn-warning offset-lg-1 col-lg-4 col-sm-12"
                onClick={postHandler("deny")}
            >
                {_("Deny")}
            </Button>
            <Button
                style={buttonMargin}
                className="btn btn-primary col-sm-12 col-lg-4 offset-lg-2 col-lg-3"
                onClick={postHandler("grant")}
            >
                {_("Install now")}
            </Button>
        </>
    );
}

Plan.propTypes = {
    plan: PropTypes.arrayOf(PropTypes.object).isRequired,
};

function Plan({ plan }) {
    return (
        <table className="table table-hover">
            <thead>
                <tr>
                    <th scope="col">{_("Action")}</th>
                    <th scope="col">{_("Name")}</th>
                    <th scope="col">{_("Current")}</th>
                    <th scope="col">{_("New")}</th>
                </tr>
            </thead>
            <tbody>
                {plan.map((item) => <PlanItem key={item.name} {...item} />)}
            </tbody>
        </table>
    );
}

PlanItem.propTypes = {
    name: PropTypes.string.isRequired,
    op: PropTypes.string.isRequired,
    new_ver: PropTypes.string,
    cur_ver: PropTypes.string,
};

function PlanItem({
    name, op, new_ver, cur_ver,
}) {
    return (
        <tr>
            <td>{op}</td>
            <td>{name}</td>
            <td>{cur_ver}</td>
            <td>{new_ver}</td>
        </tr>
    );
}
