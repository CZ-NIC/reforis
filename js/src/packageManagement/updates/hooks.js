/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import {
    API_STATE, useAlert, useAPIGet, useAPIPolling, useAPIPost,
} from "foris";
import { useCallback, useEffect, useState } from "react";

import API_URLs from "common/API";

const REFRESH_INTERVAL = 500; // milliseconds

/*
 * usePending is used to check/set if updater is running. It gets initial pending status from API.
 */
export function usePending() {
    const [pending, setPending] = useState(false);
    const [getUpdatesStatusResponse, getUpdatesStatus] = useAPIGet(API_URLs.updatesStatus);
    useEffect(() => {
        getUpdatesStatus();
    }, [getUpdatesStatus]);
    useEffect(() => {
        if (getUpdatesStatusResponse.state === API_STATE.SUCCESS) {
            setPending(getUpdatesStatusResponse.data.running);
        }
    }, [getUpdatesStatusResponse]);
    return [pending, setPending];
}

/*
 * useUpdateToApprove similar to useAPIGet(API_URLs.approvals). The getUpdateToApprove function
 * makes request only if displayApproval param is true. It makes initial request as well.
 */
export function useApprovals(displayApproval) {
    const [getApprovalsResponse, getApprovals] = useAPIGet(API_URLs.approvals);
    const updateToApprove = getApprovalsResponse.data;
    const getUpdateToApprove = useCallback(() => {
        // Request can be ignored if approval is hidden.
        if (displayApproval) {
            return getApprovals();
        }
    }, [displayApproval, getApprovals]);
    useEffect(() => {
        getUpdateToApprove();
    }, [getUpdateToApprove]);

    return [getApprovalsResponse, getApprovals, updateToApprove, getUpdateToApprove];
}

/*
 * usePendingPolling makes API polling until pending is false. It calls setPending to set result of
 * the polling.
 */
export function usePendingPolling(onSuccess, pending, setPending) {
    const [setAlert] = useAlert();

    const [checkStatusPollingResponse] = useAPIPolling(
        API_URLs.updatesStatus,
        REFRESH_INTERVAL,
        pending,
    );

    useEffect(() => {
        if (checkStatusPollingResponse.state === API_STATE.SUCCESS) {
            const isRunning = checkStatusPollingResponse.data.running;
            setPending(isRunning);
            if (!isRunning) {
                onSuccess();
            }
        } else if (checkStatusPollingResponse.state === API_STATE.ERROR) {
            setPending(false);
            setAlert(_("Cannot fetch updater status."));
        }
    },
    [checkStatusPollingResponse, onSuccess, setPending, setAlert]);
}

/*
 * useCheckUpdates is used to check for new updates. Sets pending as true when started.
 */
export function useCheckUpdates(setPending) {
    const [setAlert] = useAlert();

    const [runUpdatesResponse, runUpdates] = useAPIPost(API_URLs.runUpdates);
    // Trigger updater status checker after updater is enabled
    useEffect(() => {
        if (runUpdatesResponse.state === API_STATE.SUCCESS) {
            setPending(true);
        } else if (runUpdatesResponse.state === API_STATE.ERROR) {
            setPending(false);
            setAlert(runUpdatesResponse.data);
        }
    }, [runUpdatesResponse, setPending, setAlert]);

    return () => {
        setPending(true);
        runUpdates();
    };
}
