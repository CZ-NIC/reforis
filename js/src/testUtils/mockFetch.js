/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

export default function mockFetch(data) {
    return jest.fn(() =>
        new Promise((resolve, reject) => {
            resolve({
                ok: true,
                json: () => {
                    return data ? data : {};
                },
            });
        })
    );
}
