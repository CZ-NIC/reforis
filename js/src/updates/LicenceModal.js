/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react'

export default function LicenceModal() {
    return <div
        className="modal fade"
        id="licenceModal"
        tabIndex="-1"
        role="dialog"
        aria-labelledby="exampleModalLabel"
        aria-hidden="true"
    >
        <div className="modal-dialog" role="document">
            <div className="modal-content">
                <div className="modal-header">
                    <h5 className="modal-title">{_('Most important license agreement points')}</h5>
                    <button type="button" className="close" data-dismiss="modal" aria-label="Close">
                        <span aria-hidden="true">&times;</span>
                    </button>
                </div>
                <div className="modal-body">
                    <ul dangerouslySetInnerHTML={{
                        __html: _(`
                    <li>Automatic updates are offered to the Turris router owners free of charge.</li>
                    <li>Updates are prepared exclusively by CZ.NIC, z. s. p. o.</li>
                    <li>
                        Enabling of the automatic updates is a prerequisite for additional security features of
                        Turris router.
                    </li>
                    <li>
                        Automatic updates take place at the time of their release, the time of installation
                        cannot be influenced by the user.
                    </li>
                    <li>
                        Having the automatic updates turned on can result in increased Internet traffic on your
                        router. Expenses related to this increase are covered by you.
                    </li>
                    <li>
                        Automatic updates cannot protect you against every attack coming from the Internet.
                        Please do not forget to protect your workstations and other devices by installing
                        antivirus software and explaining to your family members how to stay safe on the
                        Internet.
                    </li>
                    <li>
                        CZ.NIC, z. s. p. o. does not guarantee the availability of this service and is not
                        responsible for any damages caused by the automatic updates.
                    </li>
                  `)
                    }}
                    />
                    <b>
                        By enabling of the automatic updates, you confirm that you are the owner of this Turris
                        router and you agree with the full text of the
                        <a href="https://www.turris.cz/omnia-updater-eula">license agreement</a>.
                    </b>
                </div>
            </div>
        </div>
    </div>
}

