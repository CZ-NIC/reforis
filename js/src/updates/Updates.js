/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react'
import moment from 'moment';

import UpdatesForm from './forms/UpdatesForm';
import ForisForm from '../formContainer/ForisForm';
import API_URLs from '../common/API';

export default function Updates({postCallback}) {
    return <ForisForm
        forisConfig={{
            endpoint: API_URLs.updates
        }}
        prepData={prepData}
        prepDataToSubmit={prepDataToSubmit}
        postCallback={postCallback}
        validator={validator}
    >
        <UpdatesForm/>
    </ForisForm>
}


function prepData(formData) {
    if (!formData.approval_settings.delay)
        formData.approval_settings.delay = 1;
    return formData
}

function prepDataToSubmit(formData) {
    if (!formData.enabled) {
        delete formData.approval_settings;
        delete formData.languages;
        delete formData.user_lists;
    } else if (formData.approval_settings.status !== 'delayed')
        delete formData.approval_settings.delay;
    return formData
}

function validator(formData) {
    let rebootErrors = {};
    if (!moment(formData.reboots.time, 'HH:mm', true).isValid())
        rebootErrors.time = _('Time should be in HH:MM format.');
    if (0 > formData.reboots.delay || formData.reboots.delay > 10)
        rebootErrors.delay = _('Number of days that must pass between receiving the request for restart and the automatic restart itself.');
    return rebootErrors.time || rebootErrors.delay ? {reboots: rebootErrors} : undefined;
}
