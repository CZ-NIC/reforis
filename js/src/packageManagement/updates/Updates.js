import React from "react";

import UpdateApprovals from "./UpdateApprovals";

export default function Updates() {
    return (
        <>
            <h1>{_("Updates")}</h1>
            <UpdateApprovals />
        </>
    );
}
