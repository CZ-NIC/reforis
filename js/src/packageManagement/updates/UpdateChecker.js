/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";

import { Button } from "foris";
import { useCheckUpdates, usePendingPolling } from "./hooks";

UpdateChecker.propTypes = {
    onSuccess: PropTypes.func.isRequired,
    pending: PropTypes.bool.isRequired,
    setPending: PropTypes.func.isRequired,
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node,
    ]),
};

export default function UpdateChecker({
    onSuccess, pending, setPending, children,
}) {
    usePendingPolling(onSuccess, pending, setPending);
    const refreshUpdates = useCheckUpdates(setPending);

    return (
        <Button
            className="btn-primary mb-3"
            forisFormSize
            onClick={refreshUpdates}
            disabled={pending}
        >
            {children}
        </Button>
    );
}
