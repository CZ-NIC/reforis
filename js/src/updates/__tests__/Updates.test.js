/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import {render, act, fireEvent, getByLabelText} from 'react-testing-library';

import {updatesFixture} from './__fixtures__/updates';
import mockFetch from '../../testUtils/mockFetch';
import Updates from '../Updates';


describe('<Updates/>', () => {
    let updatesContainer;
    beforeEach(() => {
        global.fetch = mockFetch(updatesFixture());
        const {container} = render(<Updates/>);
        updatesContainer = container;
    });

    it('Test with snapshot disabled.', () => {
        expect(updatesContainer.firstChild).toMatchSnapshot()
    });

    it('Test with snapshot enabled.', () => {
        act(() => {
            fireEvent.click(getByLabelText(updatesContainer, 'Enable automatic updates (recommended)'));
        });
        expect(updatesContainer.firstChild).toMatchSnapshot()
    });

    it('Test with snapshot delayed.', () => {
        act(() => {
            fireEvent.click(getByLabelText(updatesContainer, 'Enable automatic updates (recommended)'));
            fireEvent.click(getByLabelText(updatesContainer, 'Delayed updates'));
        });
        expect(updatesContainer.firstChild).toMatchSnapshot()
    });
});