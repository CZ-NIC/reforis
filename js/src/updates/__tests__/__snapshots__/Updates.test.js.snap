// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Updates/> Test with snapshot delayed. 1`] = `
<form>
  <div
    class="form-group row"
  >
    <div
      class="form-label col-sm-3"
    >
      <label
        class="form-label"
        for="14"
      >
        Enable automatic updates (recommended)
      </label>
    </div>
    <div
      class="form-check col-sm-8"
    >
      <input
        id="14"
        type="checkbox"
      />
      <small
        class="form-text text-muted"
      />
    </div>
  </div>
  <h4>
    Update approvals
  </h4>
  <p>
    Update approvals can be useful when you want to make sure that updates won't harm your specific
            configuration. You can e.g. install updates when you're prepared for a possible rollback to a previous
            snapshot and deny the questionable update temporarily. It isn't possible to decline the update forever an
            it will be offered to you again together with the next package installation.
  </p>
  <div
    class="form-group row"
    id="18"
  >
    <div
      class="form-label col-sm-3"
    >
      <label
        class="form-label"
        for="18"
      />
    </div>
    <div
      class="col-sm-8"
    >
      <div
        class="form-check form-check-inline"
      >
        <input
          class="form-check-input"
          id="approvals-0"
          name="approvals"
          type="radio"
          value="off"
        />
        <label
          class="form-check-label"
          for="approvals-0"
        >
          Automatic installation
        </label>
      </div>
      <small
        class="form-text text-muted"
      >
        Updates will be installed without user's intervention.
      </small>
      <div
        class="form-check form-check-inline"
      >
        <input
          class="form-check-input"
          id="approvals-1"
          name="approvals"
          type="radio"
          value="delayed"
        />
        <label
          class="form-check-label"
          for="approvals-1"
        >
          Delayed updates
        </label>
      </div>
      <small
        class="form-text text-muted"
      >
        Updates will be installed with an adjustable delay. You can also approve them manually.
      </small>
      <div
        class="form-check form-check-inline"
      >
        <input
          checked=""
          class="form-check-input"
          id="approvals-2"
          name="approvals"
          type="radio"
          value="on"
        />
        <label
          class="form-check-label"
          for="approvals-2"
        >
          Update approval needed
        </label>
      </div>
      <small
        class="form-text text-muted"
      >
        You have to approve the updates, otherwise they won't be installed.
      </small>
      <small
        class="form-text text-muted"
      />
    </div>
  </div>
  <div
    class="form-group row"
  >
    <label
      class="form-control-label col-sm-3"
      for="23"
    >
      Delayed updates (hours)
    </label>
    <div
      class="col-sm-8"
    >
      <div
        class="input-group"
      >
        <input
          class="form-control "
          id="23"
          max="168"
          min="1"
          type="number"
          value="1"
        />
      </div>
      <div
        class="invalid-feedback"
      />
      <small
        class="form-text text-muted"
      />
    </div>
  </div>
  <h4>
    Automatic restarts after software update
  </h4>
  <div
    class="form-group row"
  >
    <label
      class="form-control-label col-sm-3"
      for="15"
    >
      Delay (days)
    </label>
    <div
      class="col-sm-8"
    >
      <div
        class="input-group"
      >
        <input
          class="form-control "
          id="15"
          max="10"
          min="0"
          type="number"
          value="4"
        />
      </div>
      <div
        class="invalid-feedback"
      />
      <small
        class="form-text text-muted"
      >
        Number of days that must pass between receiving the request for restart and the automatic restart itself.
      </small>
    </div>
  </div>
  <div
    class="rdt"
  >
    <div>
      <div
        class="form-group row"
      >
        <label
          class="form-control-label col-sm-3"
          for="16"
        >
          Reboot time
        </label>
        <div
          class="col-sm-8"
        >
          <div
            class="input-group"
          >
            <input
              class="form-control "
              id="16"
              type="text"
              value="04:30"
            />
          </div>
          <div
            class="invalid-feedback"
          />
          <small
            class="form-text text-muted"
          >
            Time of day of automatic reboot in HH:MM format.
          </small>
        </div>
      </div>
    </div>
    <div
      class="rdtPicker"
    >
      <div
        class="rdtTime"
      >
        <table>
          <tbody>
            <tr>
              <td>
                <div
                  class="rdtCounters"
                >
                  <div
                    class="rdtCounter"
                  >
                    <span
                      class="rdtBtn"
                    >
                      ▲
                    </span>
                    <div
                      class="rdtCount"
                    >
                      4
                    </div>
                    <span
                      class="rdtBtn"
                    >
                      ▼
                    </span>
                  </div>
                  <div
                    class="rdtCounterSeparator"
                  >
                    :
                  </div>
                  <div
                    class="rdtCounter"
                  >
                    <span
                      class="rdtBtn"
                    >
                      ▲
                    </span>
                    <div
                      class="rdtCount"
                    >
                      30
                    </div>
                    <span
                      class="rdtBtn"
                    >
                      ▼
                    </span>
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
  <button
    class="btn btn-primary  offset-lg-8 col-lg-3 col-sm-12"
  >
     
     
    Save
  </button>
</form>
`;

exports[`<Updates/> Test with snapshot disabled. 1`] = `
<form>
  <div
    class="form-group row"
  >
    <div
      class="form-label col-sm-3"
    >
      <label
        class="form-label"
        for="1"
      >
        Enable automatic updates (recommended)
      </label>
    </div>
    <div
      class="form-check col-sm-8"
    >
      <input
        disabled=""
        id="1"
        type="checkbox"
      />
      <small
        class="form-text text-muted"
      />
    </div>
  </div>
  <h4>
    Automatic restarts after software update
  </h4>
  <div
    class="form-group row"
  >
    <label
      class="form-control-label col-sm-3"
      for="2"
    >
      Delay (days)
    </label>
    <div
      class="col-sm-8"
    >
      <div
        class="input-group"
      >
        <input
          class="form-control "
          disabled=""
          id="2"
          max="10"
          min="0"
          type="number"
          value="4"
        />
      </div>
      <div
        class="invalid-feedback"
      />
      <small
        class="form-text text-muted"
      >
        Number of days that must pass between receiving the request for restart and the automatic restart itself.
      </small>
    </div>
  </div>
  <div
    class="rdt"
  >
    <div>
      <div
        class="form-group row"
      >
        <label
          class="form-control-label col-sm-3"
          for="3"
        >
          Reboot time
        </label>
        <div
          class="col-sm-8"
        >
          <div
            class="input-group"
          >
            <input
              class="form-control "
              disabled=""
              id="3"
              type="text"
              value="04:30"
            />
          </div>
          <div
            class="invalid-feedback"
          />
          <small
            class="form-text text-muted"
          >
            Time of day of automatic reboot in HH:MM format.
          </small>
        </div>
      </div>
    </div>
    <div
      class="rdtPicker"
    >
      <div
        class="rdtTime"
      >
        <table>
          <tbody>
            <tr>
              <td>
                <div
                  class="rdtCounters"
                >
                  <div
                    class="rdtCounter"
                  >
                    <span
                      class="rdtBtn"
                    >
                      ▲
                    </span>
                    <div
                      class="rdtCount"
                    >
                      4
                    </div>
                    <span
                      class="rdtBtn"
                    >
                      ▼
                    </span>
                  </div>
                  <div
                    class="rdtCounterSeparator"
                  >
                    :
                  </div>
                  <div
                    class="rdtCounter"
                  >
                    <span
                      class="rdtBtn"
                    >
                      ▲
                    </span>
                    <div
                      class="rdtCount"
                    >
                      30
                    </div>
                    <span
                      class="rdtBtn"
                    >
                      ▼
                    </span>
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
  <button
    class="btn btn-primary  offset-lg-8 col-lg-3 col-sm-12"
    disabled=""
  >
    <span
      aria-hidden="true"
      class="spinner-border spinner-border-sm"
      role="status"
    />
     
     
     
    Restarting after %d sec.
  </button>
</form>
`;

exports[`<Updates/> Test with snapshot enabled. 1`] = `
<form>
  <div
    class="form-group row"
  >
    <div
      class="form-label col-sm-3"
    >
      <label
        class="form-label"
        for="7"
      >
        Enable automatic updates (recommended)
      </label>
    </div>
    <div
      class="form-check col-sm-8"
    >
      <input
        id="7"
        type="checkbox"
      />
      <small
        class="form-text text-muted"
      />
    </div>
  </div>
  <h4>
    Update approvals
  </h4>
  <p>
    Update approvals can be useful when you want to make sure that updates won't harm your specific
            configuration. You can e.g. install updates when you're prepared for a possible rollback to a previous
            snapshot and deny the questionable update temporarily. It isn't possible to decline the update forever an
            it will be offered to you again together with the next package installation.
  </p>
  <div
    class="form-group row"
    id="11"
  >
    <div
      class="form-label col-sm-3"
    >
      <label
        class="form-label"
        for="11"
      />
    </div>
    <div
      class="col-sm-8"
    >
      <div
        class="form-check form-check-inline"
      >
        <input
          class="form-check-input"
          id="approvals-0"
          name="approvals"
          type="radio"
          value="off"
        />
        <label
          class="form-check-label"
          for="approvals-0"
        >
          Automatic installation
        </label>
      </div>
      <small
        class="form-text text-muted"
      >
        Updates will be installed without user's intervention.
      </small>
      <div
        class="form-check form-check-inline"
      >
        <input
          class="form-check-input"
          id="approvals-1"
          name="approvals"
          type="radio"
          value="delayed"
        />
        <label
          class="form-check-label"
          for="approvals-1"
        >
          Delayed updates
        </label>
      </div>
      <small
        class="form-text text-muted"
      >
        Updates will be installed with an adjustable delay. You can also approve them manually.
      </small>
      <div
        class="form-check form-check-inline"
      >
        <input
          checked=""
          class="form-check-input"
          id="approvals-2"
          name="approvals"
          type="radio"
          value="on"
        />
        <label
          class="form-check-label"
          for="approvals-2"
        >
          Update approval needed
        </label>
      </div>
      <small
        class="form-text text-muted"
      >
        You have to approve the updates, otherwise they won't be installed.
      </small>
      <small
        class="form-text text-muted"
      />
    </div>
  </div>
  <h4>
    Automatic restarts after software update
  </h4>
  <div
    class="form-group row"
  >
    <label
      class="form-control-label col-sm-3"
      for="8"
    >
      Delay (days)
    </label>
    <div
      class="col-sm-8"
    >
      <div
        class="input-group"
      >
        <input
          class="form-control "
          id="8"
          max="10"
          min="0"
          type="number"
          value="4"
        />
      </div>
      <div
        class="invalid-feedback"
      />
      <small
        class="form-text text-muted"
      >
        Number of days that must pass between receiving the request for restart and the automatic restart itself.
      </small>
    </div>
  </div>
  <div
    class="rdt"
  >
    <div>
      <div
        class="form-group row"
      >
        <label
          class="form-control-label col-sm-3"
          for="9"
        >
          Reboot time
        </label>
        <div
          class="col-sm-8"
        >
          <div
            class="input-group"
          >
            <input
              class="form-control "
              id="9"
              type="text"
              value="04:30"
            />
          </div>
          <div
            class="invalid-feedback"
          />
          <small
            class="form-text text-muted"
          >
            Time of day of automatic reboot in HH:MM format.
          </small>
        </div>
      </div>
    </div>
    <div
      class="rdtPicker"
    >
      <div
        class="rdtTime"
      >
        <table>
          <tbody>
            <tr>
              <td>
                <div
                  class="rdtCounters"
                >
                  <div
                    class="rdtCounter"
                  >
                    <span
                      class="rdtBtn"
                    >
                      ▲
                    </span>
                    <div
                      class="rdtCount"
                    >
                      4
                    </div>
                    <span
                      class="rdtBtn"
                    >
                      ▼
                    </span>
                  </div>
                  <div
                    class="rdtCounterSeparator"
                  >
                    :
                  </div>
                  <div
                    class="rdtCounter"
                  >
                    <span
                      class="rdtBtn"
                    >
                      ▲
                    </span>
                    <div
                      class="rdtCount"
                    >
                      30
                    </div>
                    <span
                      class="rdtBtn"
                    >
                      ▼
                    </span>
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
  <button
    class="btn btn-primary  offset-lg-8 col-lg-3 col-sm-12"
  >
     
     
    Save
  </button>
</form>
`;
